macosx/Controller.mm

```diff
@@ -868,47 +868,57 @@ - (BOOL)applicationShouldHandleReopen:(NSApplication*)app hasVisibleWindows:(BOO
-            NSAlert* alert = [[NSAlert alloc] init];
-            alert.alertStyle = NSAlertStyleInformational;
-            alert.messageText = NSLocalizedString(@"Are you sure you want to quit?", "Confirm Quit panel -> title");
-            alert.informativeText = active == 1 ?
-                NSLocalizedString(
-                    @"There is an active transfer that will be paused on quit."
-                     " The transfer will automatically resume on the next launch.",
-                    "Confirm Quit panel -> message") :
-                [NSString stringWithFormat:NSLocalizedString(
-                                               @"There are %lu active transfers that will be paused on quit."
-                                                " The transfers will automatically resume on the next launch.",
-                                               "Confirm Quit panel -> message"),
-                                           active];
-            [alert addButtonWithTitle:NSLocalizedString(@"Quit", "Confirm Quit panel -> button")];
-            [alert addButtonWithTitle:NSLocalizedString(@"Cancel", "Confirm Quit panel -> button")];
+    BOOL preventedByTransfer = [self.fDefaults boolForKey:@"CheckQuitDownloading"] ? downloading > 0 : active > 0;
 
-            [alert beginSheetModalForWindow:self.fWindow completionHandler:^(NSModalResponse returnCode) {
-                [NSApp replyToApplicationShouldTerminate:returnCode == NSAlertFirstButtonReturn];
-            }];
-            return NSTerminateLater;
-        }
+    if (!preventedByTransfer)
+    {
+        return NSTerminateNow;
     }
 
-    return NSTerminateNow;
+    NSAlert* alert = [[NSAlert alloc] init];
+    alert.alertStyle = NSAlertStyleInformational;
+    alert.messageText = NSLocalizedString(@"Are you sure you want to quit?", "Confirm Quit panel -> title");
+    alert.informativeText = active == 1 ?
+        NSLocalizedString(
+            @"There is an active transfer that will be paused on quit."
+             " The transfer will automatically resume on the next launch.",
+            "Confirm Quit panel -> message") :
+        [NSString stringWithFormat:NSLocalizedString(
+                                       @"There are %lu active transfers that will be paused on quit."
+                                        " The transfers will automatically resume on the next launch.",
+                                       "Confirm Quit panel -> message"),
+                                   active];
+    [alert addButtonWithTitle:NSLocalizedString(@"Quit", "Confirm Quit panel -> button")];
+    [alert addButtonWithTitle:NSLocalizedString(@"Cancel", "Confirm Quit panel -> button")];
+    alert.showsSuppressionButton = YES;
+
+    [alert beginSheetModalForWindow:self.fWindow completionHandler:^(NSModalResponse returnCode) {
+        if (alert.suppressionButton.state == NSControlStateValueOn)
+        {
+            [self.fDefaults setBool:NO forKey:@"CheckQuit"];
```

Comment:
I might be wrong here, but won't it cause a retain cycle here? I'm thinking self -> window -> alert -> block -> self

Reply:
Is there way to test it properly?
Is there's a bug, do you know, how to fix it?
One easy fix I know is to use `NSUserDefaults.standardUserDefaults` directly here.